### PR TITLE
DAS-221 Max/Min parameter values

### PR DESCRIPTION
**Pull request by AndrewSazonov**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: Scripts/InstallPyPackages.py [hunk 1]
```diff
@@ -67,12 +67,13 @@ def install(*packages):
...
 
     installFromGit(owner='ikibalin', repo='cryspy', branch='transition-to-version-0.2')
+    installFromGit(owner='easyDiffraction', repo='easyInterface', branch='master') # Until master branch is uploaded to PIP
 
     install(
         #'cryspy==0.1.13',
         'scipy==1.4.1',
         'numpy==1.18.1',
-        'easyInterface>=0.0.6',
+        #'easyInterface>=0.0.6',
```
Comment (rozyczko) [inline Scripts/InstallPyPackages.py]: Is this supposed to be commented out? Will the master branch of easyInterface be sufficient?

File: App/PyImports/DisplayModels/FitablesModel.py [hunk 2]
```diff
@@ -10,6 +12,7 @@
 class FitablesModel(BaseModel):
     def __init__(self, parent=None):
         super().__init__(parent)
+        self._log = logger.getLogger(__class__.__module__)
```
[the remark this answers is not in the document]
Comment (wardsimon) [reply App/PyImports/DisplayModels/FitablesModel.py]: I personally find it easier to see the full path (as it is otherwise not shown in logging)

File: App/PyImports/DisplayModels/FitablesModel.py [hunk 3]
```diff
@@ -70,75 +84,139 @@ def _setModelsFromProjectDict(self):
...
         fitable_value = edit_value
+        self._log.debug(f"fitable_name: {fitable_name}")
+        self._log.debug(f"fitable_value: {fitable_value}")
+
+        if display_role_name == 'refine':
+            undo_redo_text = f"Changing '{fitable_name}' refine state to '{fitable_value}'"
+            self._log.debug(f"undo_redo_text: {undo_redo_text}")
+            self._calculator_interface.project_dict.startBulkUpdate(undo_redo_text)
+            self._calculator_interface.canUndoOrRedoChanged.emit()
+            if 'phases' == keys_list[0]:
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/DisplayModels/FitablesModel.py]: A new 'data_block_name' variable has been added and positions have been changed to if <variable> == <constant>.

File: App/PyImports/DisplayModels/FitablesModel.py [hunk 4]
```diff
@@ -70,75 +84,139 @@ def _setModelsFromProjectDict(self):
...
-                    try:
-                        self._calculator_interface.setExperimentValue(keys_list[1], keys_list[2:-2], edit_value)
-                    except AttributeError:
-                        self._calculator_interface.project_dict.setItemByPath(keys_list, edit_value)
-                        self._calculator_interface.updateExperiments()
-                    self._calculator_interface.updateCalculations() # experiments also updated ?
+                self._calculator_interface.setDictByPath(keys_list, edit_value)
+            # Update min and max if value is outside [min, max] range
+            value = self._calculator_interface.project_dict.getItemByPath(keys_list)
+            min = self._calculator_interface.project_dict.getItemByPath([*keys_list[:-1], 'min'])
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/DisplayModels/FitablesModel.py]: Fixed. See ddf4aa561e4582cafcd2032cb3f52437652cd7c4

File: App/PyImports/DisplayModels/FitablesModel.py [hunk 5]
```diff
@@ -70,75 +84,139 @@ def _setModelsFromProjectDict(self):
...
+            max = self._calculator_interface.project_dict.getItemByPath([*keys_list[:-1], 'max'])
+            self._log.debug(f"initial min: {min}, max: {max}")
+            # TODO: the code below duplicates the code from BaseClasses.py - class Base - def updateMinMax
+            # stacked changes (for GUI triggered changes)
+            if np.isclose([value], [0]):
+                min = -1
+                max = 1
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/DisplayModels/FitablesModel.py]: Done. See 22f1dd73b15c9e2719c10d41adeb8793d33c9157

File: Scripts/InstallPyPackages.py (hunk 1 above)
Comment (AndrewSazonov) [reply Scripts/InstallPyPackages.py]: Currently, min/max implementation is already pushed into master branch, but PIP package is not yet updated. So, we need this for the moment.

File: App/QmlImports/easyAnalysis/App/ContentArea/Sidebar/Pages/Analysis/Fitting/Controls.qml [hunk 6]
```diff
@@ -115,22 +115,26 @@ ColumnLayout {
         onVisibleChanged: {
             if (!visible)
                 return
-
-            const res = Specific.Variables.refinementResult
-
+            const res = Specific.Variables.refinementResult//Specific.Variables.refinementResult
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/QmlImports/easyAnalysis/App/ContentArea/Sidebar/Pages/Analysis/Fitting/Controls.qml]: Fixed. See d3979a55a3a8fe5aace66a9e34b48782eb180ed1

File: App/PyImports/DisplayModels/FitablesModel.py [hunk 7]
```diff
@@ -70,75 +84,139 @@ def _setModelsFromProjectDict(self):
...
+        Define what to do if model is changed, e.g. from GUI.
+        """
         role = roles[0]
         role_name = self._roles_dict[role].decode()
+        self._log.debug(f"roles: {roles}")
+        self._log.debug(f"role: {role}")
+        self._log.debug(f"role_name: {role_name}")
+
         if role_name.endswith(self._edit_role_name_suffix):
             index = top_left_index
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/DisplayModels/FitablesModel.py]: Done. See 138ba3e1a8465792407354085205247568ac5a2b

File: App/PyImports/DisplayModels/FitablesModel.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/DisplayModels/FitablesModel.py]: I also prefer to see the full path

File: App/PyImports/DisplayModels/FitablesModel.py [hunk 8]
```diff
@@ -70,75 +84,139 @@ def _setModelsFromProjectDict(self):
...
+            self._calculator_interface.project_dict.startBulkUpdate(undo_redo_text)
+            self._calculator_interface.canUndoOrRedoChanged.emit()
+            if 'phases' == keys_list[0]:
+                try:
+                    self._calculator_interface.setPhaseRefine(keys_list[1], keys_list[2:-2], edit_value)
+                except AttributeError:
+                    # In this case the calculator/dict are out of phase :-/ So fallback to manual.
+                    self._calculator_interface.project_dict.setItemByPath(keys_list, edit_value)
+                    # Sync and update so this shouldn't happen again.
+                #self._calculator_interface.updatePhases()
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/DisplayModels/FitablesModel.py]: Comment is deleted. See 06ac2438cd81a9c7e2fff48ba2f6680cf92a09b1

File: App/PyImports/Utils/QtLogging.py [hunk 9]
```diff
@@ -22,6 +22,15 @@ def __init__(self, logger: Logger, parent=None):
             { 'level': { 'name': 'Error',    'code': logging.ERROR } },
             { 'level': { 'name': 'Critical', 'code': logging.CRITICAL } }
         ]
+        self.setLevel(self.defaultLevelIndex())
+
+    def initialLevel(self, level):
+        """
+        ...
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/Utils/QtLogging.py]: This method is needed to set the initial 'qt logger level (QtLogger)' to be different from 'base logger level (Logger)' when creating an object of QtLogger class, i.e. when 'level' is explicitly defined. If 'level' is not explicitly defined, then QtLogger object takes the logging level from Logger object.

**Comment by AndrewSazonov** [comment]
> wouldn't it make more sense to rename min/max in lines 173/174?

It would, my fault. Now fixed. See f28f075c3fe5d36fffae0348092e53c8085adfed

File: App/PyImports/Utils/Helpers.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+                children.append(xml)
+
+    else:
+        #if list
+        for value in d:
+            children.append(dict2xml(value, root_singular))
+
+    end_tag = '>' if 0 < len(children) else '/>'
+
+    if wrap or isinstance(d, dict):
```
Comment (rozyczko) [inline App/PyImports/Utils/Helpers.py]: `wrap` already contains conditional for isinstance(d, dict)

File: App/PyImports/Utils/Helpers.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
+def dict2xml(d, root_node=None):
+    wrap          =  False if None == root_node or isinstance(d, list) else True
+    root          = 'root' if None == root_node else root_node
+    root_singular = root[:-1] if 's' == root[-1] and None == root_node else root
```
Comment (rozyczko) [inline App/PyImports/Utils/Helpers.py]: if `root_node == None`, your `root` is already set to `"root"`, so the first part of the conditional in line 4 is redundant!

File: App/PyImports/Utils/Helpers.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+    end_tag = '>' if 0 < len(children) else '/>'
+
+    if wrap or isinstance(d, dict):
+        xml = '<' + root + attr + end_tag
+
+    if 0 < len(children):
+        for child in children:
+            xml = xml + child
+
```
Comment (rozyczko) [inline App/PyImports/Utils/Helpers.py]: or list comprehension
`xml = [i for i in children]`

**Comment by rozyczko** [comment]
Please consider standardizing on `if <variable> <operand> <constant>` syntax - it's much more common and easier to read. E.g.
`root  = 'root' if root_node is None else root_node` instead of
`root  = 'root' if None == root_node else root_node`

File: App/PyImports/Utils/Helpers.py (hunk 11 above)
Comment (rozyczko) [reply App/PyImports/Utils/Helpers.py]: Also, please don't compare `None` with `==` - use `is` instead.
`if root_node is None` rather than `if root_node == None`
In general, use `is` when you want to check against an object's identity (e.g. checking to see if var is None). Use `==` when you want to check equality (e.g. Is var equal to 3?).
Your object you are comparing to `None` can have `__eq__` redefined thus rendering the conditional not obvious.

File: App/PyImports/Utils/Helpers.py (hunk 10 above)
Comment (AndrewSazonov) [reply App/PyImports/Utils/Helpers.py]: `wrap` contains` isinstance(d, list)`, not `isinstance(d, dict)`

File: App/PyImports/Utils/Helpers.py (hunk 11 above)
Comment (AndrewSazonov) [reply App/PyImports/Utils/Helpers.py]: Fixed. See dc4f8fced1eb47e179721c588f70737374027df3

File: App/PyImports/Utils/Helpers.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+            else:
+                xml = '<' + key + ">" + str(value) + '</' + key + '>'
+                children.append(xml)
+
+    else:
+        #if list
+        for value in d:
+            children.append(dict2xml(value, root_singular))
+
+    end_tag = '>' if 0 < len(children) else '/>'
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/Utils/Helpers.py]: Fixed. See dc4f8fced1eb47e179721c588f70737374027df3

File: App/PyImports/Utils/Helpers.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+    if wrap or isinstance(d, dict):
+        xml = '<' + root + attr + end_tag
+
+    if 0 < len(children):
```
[the remark this answers is not in the document]
Comment (AndrewSazonov) [reply App/PyImports/Utils/Helpers.py]: Fixed. See dc4f8fced1eb47e179721c588f70737374027df3

**Comment by AndrewSazonov** [comment]
> Please add tests for the other case where you pass a list, and root_node is none.

Done

File: App/PyImports/Utils/Helpers.py (hunk 12 above)
Comment (AndrewSazonov) [reply App/PyImports/Utils/Helpers.py]: Done. See 4e60380dfb36f36e53e9c24e17f93143602c5af5
